src/interfaces/setup_interface.jl

```diff
@@ -0,0 +1,171 @@
+    This behaviour needs to be overwritten, if this is not the case.
+
+"""
+@inline function _input_validation(stp::AbstractComputationSetup, input)
+    return true
+end
+
+"""
+    
+    function _post_computation(stp::AbstractComputationSetup, input::Any, result::Any)
+    
+Interface functions which is called in [`compute`](@ref) after [`_compute`](@ref) was called. This function is dedicated to 
+finalize the result of a computation. 
+
+!!! note "default implementation"
+
+    Since in the case of no post computation, the result of [`_compute`](@ref) is not changed, this function returns `result` per default.
+
+"""
+@inline function _post_computation(stp::AbstractComputationSetup, input, result)
+    return result
+end
+
+"""
+    
+    _compute(stp::AbstractComputationSetup, input::Any)
+
+Interface function which returns the value of the associated quantity evaluated on `input`, which can be anything the associated quantity is defined to be feasable for.
+
+!!! note "unsafe implementation"
+
+    This function should be implemented without any input validation or post computations (see [`_input_validation`](@ref) and [`_post_computation`](@ref)). The latter two are performed while calling 
+    the safe version of this function [`compute`](@ref).
+
+"""
+function _compute end
+
+"""
+
+    compute(stp::AbstractComputationSetup, input::Any)
+
+Return the value of the quantity associated with `stp` for a given `input`. 
+In addition to the actual call of the associated unsafe version [`_compute`](@ref),
+input validation (using [`_input_validation`](@ref)) and post computation
+(using [`_post_computation`](@ref)) are wrapped around the calculation (see [`AbstractComputationSetup`](@ref) for details).
+"""
+function compute(stp::AbstractComputationSetup, input)
+    _input_validation(stp,input) || error("InvalidInputError: there is something wrong with the input!\n setup:$stp \n input:$input")
+    raw_result = _compute(stp,input)
+    return _post_computation(stp, input,raw_result)
+end
+
+"""
+Abstract base type for setups related to a combination scattering processes and compute models.  
+Every subtype of `AbstractProcessSetup` is expected to implement at least the following 
+interface functions
+
+```Julia
+scattering_process(::AbstractProcessSetup) 
+compute_model(::AbstractProcessSetup) 
+```
+
+Derived from the these interface functions, the following delegations are implemented
+
+```Julia
+number_incoming_particles(::AbstractProcessSetup)
+number_outgoing_particles(::AbstractProcessSetup)
+```
+
+"""
+abstract type AbstractProcessSetup <: AbstractComputationSetup end
+
+"""
+    
+    scattering_process(stp::AbstractProcessSetup)
+
+Interface function, which returns the scattering process associated with `stp`,
+i.e. an object which is a subtype of `AbstractProcessDefinition`. 
+"""
+function scattering_process end
```

Comment:
As with compute, I'd at least add the parameters and a throw with an error message:
```suggestion
function scattering_process(stp::AbstractProcessSetup)
    throw("UnimplementedError: you must implement scattering_process(stp::ProcessSetup) for every subtype of AbstractProcessSetup"
end
```